DM2C/code/model.py

```diff
@@ -0,0 +1,367 @@
+                    txt_D_loss = (self.adv_loss_fn(self.D_txt(txt_embed.detach()), txt_real) +
+                                  self.adv_loss_fn(self.D_txt(img2txt_recon.detach()), img_fake)) / 2
+                    D_loss = (img_D_loss + txt_D_loss) * self.args.lamda3
+                elif self.args.gan_type == 'wasserstein':
+                    img_D_loss = self.D_img(txt2img_recon.detach()).mean() - \
+                                 self.D_img(img_embed.detach()).mean()
+                    txt_D_loss = self.D_txt(img2txt_recon.detach()).mean() - \
+                                 self.D_txt(txt_embed.detach()).mean()
+                    D_loss = (img_D_loss + txt_D_loss) * self.args.lamda3
+                D_loss.backward()
+                self.optimizer_D.step()
+
+                # weight clipping
+                if self.args.gan_type == 'wasserstein':
+                    for p in self.D_img.parameters():
+                        p.data.clamp_(-self.args.clip_value,
+                                      self.args.clip_value)
+                    for p in self.D_txt.parameters():
+                        p.data.clamp_(-self.args.clip_value,
+                                      self.args.clip_value)
+
+            if (step + 1) % self.args.log_freq == 0:
+                self.logger.info(info_string1 % (
+                    epoch, self.args.n_epochs, step, len(self.train_loader),
+                    D_loss.item(), img_D_loss.item(), txt_D_loss.item(),
+                    G_loss.item(),  img_cycle_loss.item(),
+                    txt_cycle_loss.item()))
+                self.writer.add_scalar(
+                    'Train/G_loss', G_loss.item(),
+                    step + len(self.train_loader) * epoch)
+                self.writer.add_scalar(
+                    'Train/D_loss', D_loss.item(),
+                    step + len(self.train_loader) * epoch)
+
+        if epoch > 10 and (epoch + 1) % self.args.save_freq == 0:
+            self.save_cpt(epoch)
+
+    def _build_masterpraktikum_dataloader(self):
+        kwargs = {'num_workers': self.args.n_cpu, 'pin_memory': True}
+
+        h5ad_dataset = h5ad_Dataset(self.args.h5ad_data)
+        img_dataset = img_Dataset(self.args.img_path)
+
+        # we need to embed the h5ad data first so we can input them into the dataloader
+        h5ad_embed = torch.stack([embed for embed in self.cell_plm.calc_embed(h5ad_dataset.data)])
+        # same with images -- embed first
+        img_loader = DataLoader(dataset = img_dataset, # image loader so that not all images are read into memory for embedding calculation
+                                batch_size= self.args.batch_size,
+                                shuffle = True)
+        img_embed=[]
+        for load in img_loader:
+            img_embed.extend(self.vis_trans.calc_embed(load))
+        img_embed = torch.stack(img_embed)
```

Comment:
Are you sure your code works as expected? Note that `torch.stack` creates a new dimension (the batch dimension), but the tensors returned by `self.vis_trans.calc_embed` have already been stacked! What is the shape of img_embed? It has to be (n_images, hidden_size).

Reply:
it works but there may be a better way to do it. maybe with concatenating